### PR TITLE
FIX: asyncio server subscription-related failures and (some) test suite instability

### LINKED ISSUE #830: caproto IOCs on Windows Python 3.12 have issues
Windows Python 3.12 failure is legitimate and locally reproducible (as well as on CI).

Copied from my comment in https://github.com/caproto/caproto/pull/829:
* The accuracy of the claim below is not guaranteed - it was based on some debugging efforts which I ultimately gave up on

_How_ it's failing is a little bizarre.
```
caproto/tests/test_threading_client.py::test_unsubscribe_all[caproto]
```
This test:
* Makes 2 subscriptions of different data types to the same PV.
* Writes a few values
* Double-checks that we got the number of subscriptions expected

On Windows Python 3.12, the server only sends back the _first_ data type that was requested. This is not the case on other platforms.

I'm at a loss as to how it's happening (TODO look into this again at some point):
* The two subscription events (`EventAddResponse`) are put in as weakrefs into to the queue (`circuit.subscription_queue.put`)
* The weakrefs are not destroyed (checked via logs; they have a hard reference in "unexpired updates")
* Only *one* subscription makes it out of the queue (`ref = await self.get_from_sub_queue(...)` -> `circuit.subscription_queue.get()` wrapped with `ensure_future` to add timeout support)

<details>
<summary>Log</summary>

```
DEBUG    caproto.circ:_circuit.py:177 EventAddRequest(data_type=<ChannelType.INT: 1>, data_count=0, sid=0, subscriptionid=0, low=0.0, high=0.0, to=0.0, mask=13)
DEBUG    caproto.circ:_circuit.py:177 EventAddRequest(data_type=<ChannelType.STRING: 0>, data_count=0, sid=0, subscriptionid=1, low=0.0, high=0.0, to=0.0, mask=13)

Initial sub value for INT only comes back:

DEBUG    caproto.ch:client.py:2185 'a17b46ee:int': EventAddResponse(data=array([96], dtype='>i2'), data_type=<ChannelType.INT: 1>, data_count=1, status=CAStatusCode(name='ECA_NORMAL', code=0, code_with_severity=1, severity=<CASeverity.SUCCESS: 1>, success=1, defunct=False, description='Normal successful completion'), subscriptionid=0, metadata=None)
DEBUG    caproto.circ:client.py:1484 EventAddResponse(data=array([96], dtype='>i2'), data_type=<ChannelType.INT: 1>, data_count=1, status=CAStatusCode(name='ECA_NORMAL', code=0, code_with_severity=1, severity=<CASeverity.SUCCESS: 1>, success=1, defunct=False, description='Normal successful completion'), subscriptionid=0, metadata=None)

Write 123:

DEBUG    caproto.circ:_circuit.py:177 WriteNotifyRequest(data=array([123], dtype='>i4'), data_type=<ChannelType.LONG: 5>, data_count=1, sid=0, ioid=0, metadata=None)
DEBUG    caproto.circ:client.py:1484 WriteNotifyResponse(data_type=<ChannelType.LONG: 5>, data_count=1, status=CAStatusCode(name='ECA_NORMAL', code=0, code_with_severity=1, severity=<CASeverity.SUCCESS: 1>, success=1, defunct=False, description='Normal successful completion'), ioid=0)

See INT monitor event but not string one:

DEBUG    caproto.circ:_circuit.py:177 WriteNotifyRequest(data=array([456], dtype='>i4'), data_type=<ChannelType.LONG: 5>, data_count=1, sid=0, ioid=1, metadata=None)
DEBUG    caproto.ch:client.py:2185 'a17b46ee:int': EventAddResponse(data=array([123], dtype='>i2'), data_type=<ChannelType.INT: 1>, data_count=1, status=CAStatusCode(name='ECA_NORMAL', code=0, code_with_severity=1, severity=<CASeverity.SUCCESS: 1>, success=1, defunct=False, description='Normal successful completion'), subscriptionid=0, metadata=None)
DEBUG    caproto.circ:client.py:1484 EventAddResponse(data=array([123], dtype='>i2'), data_type=<ChannelType.INT: 1>, data_count=1, status=CAStatusCode(name='ECA_NORMAL', code=0, code_with_severity=1, severity=<CASeverity.SUCCESS: 1>, success=1, defunct=False, description='Normal successful completion'), subscriptionid=0, metadata=None)
DEBUG    caproto.circ:client.py:1484 WriteNotifyResponse(data_type=<ChannelType.LONG: 5>, data_count=1, status=CAStatusCode(name='ECA_NORMAL', code=0, code_with_severity=1, severity=<CASeverity.SUCCESS: 1>, success=1, defunct=False, description='Normal successful completion'), ioid=1)


DEBUG    caproto.ch:client.py:2185 'a17b46ee:int': EventAddResponse(data=array([456], dtype='>i2'), data_type=<ChannelType.INT: 1>, data_count=1, status=CAStatusCode(name='ECA_NORMAL', code=0, code_with_severity=1, severity=<CASeverity.SUCCESS: 1>, success=1, defunct=False, description='Normal successful completion'), subscriptionid=0, metadata=None)
DEBUG    caproto.circ:client.py:1484 EventAddResponse(data=array([456], dtype='>i2'), data_type=<ChannelType.INT: 1>, data_count=1, status=CAStatusCode(name='ECA_NORMAL', code=0, code_with_severity=1, severity=<CASeverity.SUCCESS: 1>, success=1, defunct=False, description='Normal successful completion'), subscriptionid=0, metadata=None)
```
<details>

### PR DESCRIPTION
**Pull request by klauer**
### Related issues/PRs
Closes #830
Closes #434 (9f52809b2d2b38c96622d9833220b387c49d2038)
Supersedes #845

### Changes

* Picks up #845 because curio/trio are not fun to work with in 2024
* Fixes an apparently massive bug in server subscriptions causing test instability
   * Debugging led me to [this](https://github.com/caproto/caproto/issues/830#issuecomment-2525303112), and it seems `ensure_future` was the culprit causing the queue items to be dropped when a timeout happened
   * That change is [e75cf00](https://github.com/caproto/caproto/pull/854/commits/e75cf005cbbadffbd520e6ab8b53eb7fbcd75117)
   * Fixes the worst hit Windows IOCs on 3.12, but also improves stability in all other platforms and versions
* Fixes other test suite bugs, primarily `auto_monitor=True`-related race conditions that we should have caught ages ago
* Fixes documentation building
* Fixes Azure tests

### Original text

Anecdotally, this one test I focused on runs on my home Windows machine without failing now:

```
caproto/tests/test_threading_client.py::test_unsubscribe_all[caproto] PASSED  
```

(It's not due to the tweaked `time.sleep` bits - the subscriptions are actually now making it through from the queue to the socket rather than getting dropped)

### CODE REVIEW

**Comment by klauer** [comment]
This looks good on the initial CI run - it fixes the threading client tests.

This test remains flaky even on MacOS - but it's much more so on Windows. It solely caused all of the Windows test suite failures in this first run.
```
FAILED caproto/tests/test_records.py::test_alarms[AlarmSeverity.NO_ALARM-LLSV] - assert 0 == 2
FAILED caproto/tests/test_records.py::test_alarms[AlarmSeverity.MINOR_ALARM-LLSV] - assert 1 == 2
FAILED caproto/tests/test_records.py::test_alarms[AlarmSeverity.INVALID_ALARM-LLSV] - assert 3 == 2
```

**Comment by klauer** [comment]
OK, `test_alarms` may not have a race condition but have some actual bug. Can't yet reproduce locally...

**Comment by klauer** [comment]
That was it

**Comment by klauer** [comment]
@tacaswell The commits up to [d0a49be](https://github.com/caproto/caproto/pull/854/commits/d0a49befb2845fd2316080dbb59af2c651c97263) appear to completely fix our pytest workflows on GitHub Actions on all platforms:
image

This excludes documentation and Azure Pipelines which are both failing miserably.

I may push a couple more commits for Azure but I could just as well put those in a separate PR. Review whenever you can and I'll merge whatever's here at that point, as it's still miles better than the master branch's success rate.

**Comment by klauer** [comment]
All green? Yeah, that's not going to happen twice in a row

I'm done modifying for now

**Comment by klauer** [comment]
This is a pretty big fix and yet a very simple one to review, considering the actual fix is on two lines.

Any chance you guys will find a few minutes to take a look? 

🦗 🦗 🦗

File: caproto/threading/pyepics_compat.py [hunk 1]
```diff
@@ -962,29 +962,27 @@ def disconnect(self):
 
 
 def get_pv(pvname, *args, context=None, connect=False, timeout=5, **kwargs):
-    if context is None:
```
Comment (tacaswell) [inline caproto/threading/pyepics_compat.py:965]: Thank you for fixing this.  It was on my mental to-do list but never made it up to being done!

**Pull request merged**